### ISSUE TITLE
cell magic syntax highlighting (Ruby)

### ISSUE THREAD

**Issue by MatthewSteen** [jupyter/notebook#2102]
I'm seeing this in Jupyter for 2.7. See [this comment](https://github.com/jupyter/notebook/issues/1439#issuecomment-276402940) and the following one by @takluyver.

* jupyter_client-4.4.0
* jupyter_console-5.0.0
* jupyter_core-4.2.0

---

**Jupyter - long cell magic applies Python highlighting to Ruby (expected)**

![image](https://cloud.githubusercontent.com/assets/8397034/22472054/af82af12-e791-11e6-9ad0-c8a4a9fe2caa.png)
 
---

**Jupyter - short cell magic highlights all code as a string (unexpected)**

![image](https://cloud.githubusercontent.com/assets/8397034/22472235/453a7526-e792-11e6-9ea0-bf576128f295.png)

**Comment by takluyver** [comment]
What version of the `notebook` package do you have? If you've got time, can you try installing from master (see [setting up a development environment](http://jupyter-notebook.readthedocs.io/en/latest/contributing.html#setting-up-a-development-environment))?

I think it's [this code](https://github.com/jupyter/notebook/blob/2eb7bf19530252247dc9b112016080733b150352/notebook/static/notebook/js/cell.js#L634) that is going wrong.

**Comment by MatthewSteen** [comment]
I have `notebook-4.2.3-py27_0`.

**Comment by takluyver** [comment]
Can you try with the latest version, currently 4.3.1?

**Comment by MatthewSteen** [comment]
Same result with `notebook-4.3.1-py27_1` after updating, shutting down my running notebooks, and restarting Jupyter.

**Comment by takluyver** [comment]
Thanks. In fact, I already tested on master but forgot about it earlier (sorry about that).

**Comment by Carreau** [comment]
> I think it's this code that is going wrong.

Yes, I'm almost sure of what's wrong. It switches language in the middle of the first line instead of waiting to end of first line to switch, which screw up codemirror highlighting. 

In general the auto-highlight of IPython should be folded into a **real** IPython codemirror mode. Though that's a non-negligible task. When I was familiar with CodeMirror internals and mode writing, I would have said that a _clean_ IPython mode would take 1 to 2 good uninterrupted days of work. Plus about 1, to integrate into the current notebook JS.

So I'm going to guess this will likely end up being bumped to 5.1.

**Comment by takluyver** [comment]
Bumped for now; if anyone works out a fix, I'm happy to land it for 5.0, but I don't want to hold up the release for it.

**Comment by gnestor** [comment]
@Carreau Is there a quick-and-dirty fix for this that we can get into 5.1? Otherwise, let's mark as backlog...

**Comment by Carreau** [comment]
> @Carreau Is there a quick-and-dirty fix for this that we can get into 5.1? Otherwise, let's mark as backlog...

No, not that I know of.  We might also want to push that to jupyterlab

**Comment by gnestor** [comment]
Is it [this file](https://github.com/jupyter/notebook/blob/master/notebook/static/notebook/js/codemirror-ipython.js) in notebook and [this file](https://github.com/jupyterlab/jupyterlab/blob/master/packages/codemirror/src/codemirror-ipython.ts) in jupyterlab that would need to be updated? If so, they're pretty much identical, so whatever improvements made in jupyterlab should be easy to port to notebook...
